ci/glci/openstack_image.py

```diff
@@ -0,0 +1,82 @@
+import os
+import time
+import argparse
+
+from glci.model import OpenRC
+
+from openstack import connect
+
+class OpenstackImageUploader:
+  '''OpenstackImageUploader is a client to upload images to Openstack Glance.'''
+
+  def __init__(self, openrc: OpenRC):
+      self.conn = connect(
+          auth_url=openrc.auth_url,
+          project_name=openrc.project_name,
+          username=openrc.username,
+          password=openrc.password,
+          region_name=openrc.region,
+          user_domain_name=openrc.domain,
+          project_domain_name=openrc.domain,
+      )
+
+  def upload_image(self, name: str, path: str, meta: dict, timeout_seconds=86400):
+      '''Upload an image from filesystem to Openstack Glance.'''
+
+      image = self.conn.image.create_image(
+          name=name,
+          filename=path,
+          disk_format='vmdk',
+          container_format='bare',
+          visibility='community',
+          meta=meta,
+          timeout=timeout_seconds,
+      )
+      return image['id']
+
+  def wait_image_ready(self, image_id: str, wait_interval_seconds=10):
+      '''Wait until an image get in ready state.'''
+
+      while True:
+          image = self.conn.image.get_image(image_id)
+          if image['status'] == 'queued' or image['status'] == 'saving':
+              time.sleep(wait_interval_seconds)
+              continue
+          if image['status'] == 'active':
+              return
+          RuntimeError(f"image upload to Glance failed due to image status {image['status']}")
+
+
+def upload_image(image_name: str, image_path: str, build_number: str, openrc: OpenRC, architecture='x86_64'):
```

Comment:
use `glci.model.Architecture`; translate it to OpenStack's naming scheme. This has the advantage of implicitly validating allowed / known values, and avoid code duplication (with the exception of the translation code, of course)

Reply:
obsolete.